### PR TITLE
[TT-621] fix limit inheritance & query depth limiter

### PR DESCRIPTION
**Pull request by spetrunin**
log correct limit in per field depth limit

## Related Issue
https://tyktech.atlassian.net/browse/TT-621

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [ ] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
  - [ ] If new config option added, ensure that it can be set via ENV variable
- [ ] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [ ] When updating library version must provide reason/explanation for this update.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Check your code additions will not fail linting checks:
  - [ ] `go fmt -s`
  - [ ] `go vet`

### CODE REVIEW

**Review by matiasinsaurralde — APPROVED** [review]

**Review by gernest — CHANGES_REQUESTED** [review]
Looks good. Is it possible to have a test for this? We don't want this to be messed up in the future.

**Review by sredxny — APPROVED** [review]
LGTM

File: gateway/session_manager.go [hunk 1]
```diff
@@ -188,6 +190,12 @@ func (l *SessionLimiter) ForwardMessage(r *http.Request, currentSession *user.Se
 		return sessionFailRateLimit
 	}
 
+	// If MaxQueryDepth is -1 or 0, it means unlimited and no need for depth limiting.
+	if l.DepthLimitEnabled(api.GraphQL.Enabled, accessDef) {
```
Comment (pvormste) [inline gateway/session_manager.go:194]: This code has been moved in a past PR. It now lives in `gateway/mw_graphql_complexity.go`

File: gateway/session_manager.go [hunk 2]
```diff
@@ -331,21 +339,92 @@ func GetAccessDefinitionByAPIIDOrSession(currentSession *user.SessionState, api
...
+			ThrottleInterval:   currentSession.ThrottleInterval,
+			ThrottleRetryLimit: currentSession.ThrottleRetryLimit,
+			MaxQueryDepth:      currentSession.MaxQueryDepth,
 		}
 	}
 
 	return accessDef, allowanceScope, nil
 }
+
+func (l *SessionLimiter) DepthLimitEnabled(graphqlEnabled bool, accessDef *user.AccessDefinition) bool {
```
Comment (pvormste) [inline gateway/session_manager.go:358]: Now living in: `gateway/mw_graphql_complexity.go`

File: gateway/session_manager.go [hunk 3]
```diff
@@ -331,21 +339,92 @@ func GetAccessDefinitionByAPIIDOrSession(currentSession *user.SessionState, api
...
+	// There is a possibility that depth limit is disabled on field level too,
+	// but we continue with this as we do not know per root field level values until we have calculated request complexity
+	if len(accessDef.FieldAccessRights) > 0 {
+		return true
+	}
+
+	return accessDef.Limit.MaxQueryDepth > 0
+}
+
+func (l *SessionLimiter) DepthLimitExceeded(gqlRequest *graphql.Request, accessDef *user.AccessDefinition, schema *graphql.Schema) sessionFailReason {
```
Comment (pvormste) [inline gateway/session_manager.go:372]: This also is located in `gateway/mw_graphql_complexity.go`

File: gateway/session_manager_test.go [hunk 4]
```diff
@@ -134,3 +137,255 @@ func TestGetAccessDefinitionByAPIIDOrSession(t *testing.T) {
 		assert.NoError(t, err)
 	})
 }
+
+func TestSessionLimiter_DepthLimitEnabled(t *testing.T) {
```
Comment (pvormste) [inline gateway/session_manager_test.go:141]: Depth Limit tests are now located in: `gateway/mw_graphql_test.go`

File: gateway/session_manager.go [hunk 5]
```diff
@@ -94,6 +95,7 @@ const (
 	sessionFailNone sessionFailReason = iota
 	sessionFailRateLimit
 	sessionFailQuota
+	sessionFailDepthLimit
```
Comment (pvormste) [inline gateway/session_manager.go:98]: I think this is not needed anymore

**Review by pvormste — CHANGES_REQUESTED** [review]
If you look at this PR: https://github.com/TykTechnologies/tyk/pull/3469/files you will see, that some of the code changed in the past and this PR would revert the changes in current state.

So you will need to adjust the code (tests are also failing because of this reason)

**Review by pvormste — APPROVED** [review]
Nice, all tests passing now! Good job

**Review by urbanishimwe — APPROVED** [review]

**Review by gernest — APPROVED** [review]

**Comment by tbuchaillot** [comment]
@spetrunin this PR was missing in `release-3.2` and `release-3.2.2` branches.

**Comment by sredxny** [comment]
@spetrunin there're some issues merging this into release-3.2 and release-3.2.2, could you help us to include these changes?

**Pull request merged**